### ISSUE TITLE
Server-side (DataPortal Host) caching of business objects

### ISSUE THREAD

**Issue by oatsoda** [MarimerLLC/cslaforum#30]
I'm looking to add some Business Object caching on the server side of our application which is currently using a 2-tier model with the server-side hosted within Azure.

I've been reading various forum posts and also looking at the CslaContrib repo but nothing seems to suit my needs.  The CslaContrib appears to be client-side caching (Although I may be wrong but CacheDataPortal seems to be implementing IDataPortalProxy - I suppose greater than 2-tier might be able to utilise this)

Ideally what I would like to do is intercept calls on the server side, after IAuthorizeDataPortal, and implement a caching layer (to utilise an Azure Redis cache) but there appears to be no way to do this.

From what I can see my only options are:

1. Implement my own Hosts.IWcfPortal (or subclass Hosts.WcfPortal) ensuring I make the IAuthorizeDataPortal myself and then delegating off to a caching layer which eventually would call Server.DataPortal (as per Hosts.WcfPortal does) when there is a cache-miss.

or

2. Implement DataPortal_OnDataPortalInvoke (or similar) on each business object (or on a base) and copy over cached data from a cached BO the currently instantiated "shell" object.

or

3. Implement caching at a Data Access level instead.

Does anyone have any suggestions for how I can achieve what I would like to do?

**Comment by jonnybee** [comment]
Hi,
Which data portal implementation do you use in your app? 

I prefer to use ObjectFactory method and then the object factory class is resposible for creating the instance and return to client. IE: Could also deserialize an object from a cache and return. 

If you use DataPortal_XYZ method the internal DataPortal implementation will create the BO and the your data access code is instance methods. 

Also look ant the ObjectFactoryLoader. This is effectively an interception after IAuthorizeDataPortal where you can also use IoC to load the ObjectFactory class. ObjectFactoryLoader is reposible for returning the class that implements the data access methods for a "root" BO. 

My gut feeling is that this is much easier to implement with ObjectFactory approach.

**Comment by oatsoda** [comment]
Hi Jonny

I am using the standard DataPortal_ implementation, not the ObjectFactory.  I was always put off with the ObjectFactory requiring the reflected name of the factory type.  But is this still required if I use the ObjectFactoryLoader?

**Comment by jonnybee** [comment]
No, you can even use interface in ObjectFactory declaration and your own choice of IoC container in ObjectFactoryLoader. 

Sample using MEF: 

      [Serializable]
      [ObjectFactory(typeof(IMyRootFactory))]
      public class MyRoot : BusinessBase<MyRoot>

and implementation: 

    [Export(typeof(IMyRootFactory))]
    public class MyRootFactory : ObjectFactory, IMyRootFactory
    {
        [RunLocal]
        public object Create()
        {
             ......
    	}

        public object Fetch(object criteria)
        {
             ......
        }

For more code see CslaContrib and $\Samples\MEFSamples

**Comment by oatsoda** [comment]
Thanks Jonny.  I've had a look through that - unfortunately I am using Ninject for IoC, so Export not really available.  But then I suppose I am looking to implement object factory for caching, not for abstracting away my data access like in the examples.

This has given me something to build on so thanks very much.

**Comment by jonnybee** [comment]
That just means that you need to implement an ObjectFactoryLoader that use NInject as IoC container. 

The sample used MEF just because MEF is standard in .NET. 

You must implement some object factory to handle data access. I haven't tried to create a ObjectFactory that would create the object and then call the DataPortal_XYZ methods similar to DataPortal (or even looked into whether that would be possible). 

Keep us posted on your progress!

**Comment by oatsoda** [comment]
I suppose if I create an ObjectFactoryLoader and make sure that my IoC kernel is statically set on the type then when CSLA creates it I can have access to the Kernel to resolve.

That is basically what I am thinking - my ObjectFactory becomes a proxy which calls the dataportal...but I don' want to completey reproduce the logic in the [DataPortal](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/Server/DataPortal.cs)!

Do you know why in that DataPortal the `GetObjectFactoryAttribute` and `GetFactoryType` are only implemented in the `Update` and `Delete`?

**Comment by oatsoda** [comment]
Ok - Not sure exactly why `Update` and `Delete` have additional logic inside `DataPortal`, but I suppose these are the methods that "change" data (as opposed to Fetch, Create) so it is probably due to that....but I think this is irrelevant.

There are three DataPortal implementations:

TransactionalDataPortal - for wrapping <x>DataPortal calls in transactions
FactoryDataPortal - for the ObjectFactory mechanism
SimpleDataPortal - for the encapsulated DataPortal_<x> mechanism

By the time the FactoryDataPortal is executed we should already be within a Transaction if one was required, so I *think* I am correct in saying that I could fall back to the SimpleDataPortal for regular operation.

The only other thing I have to find out is how the ChildDataPortal comes into play - I know these calls don't go over-the-wire but I guess if the Child type is marked with the ObjectFactory attribute it will still invoke the object factory...so I still need to figure this part out.

**Comment by jonnybee** [comment]
ChildDataPortal does NOT come into play. There is no support for ObjectFactory attribute on child objects (only the "root" DataPortal methods will use the ObjectFactory attribute). 

If you want to use transaction you must apply TracactionAttribute to the method that is called from DataPortal in the ObjectFactory class. See http://www.lhotka.net/weblog/CSLANETObjectFactoryAttribute.aspx

**Comment by oatsoda** [comment]
Hmm, this kind of brings another question into play...the Child would therefore be cached as part of the Parent object graph.  But what if this Child was used by another Parent?  Or what if this Child also could be used as a Root object...

**Comment by rockfordlhotka** [comment]
A child can only be referenced by a single parent. CSLA works under the assumption that an object graph is self-contained. If you create a reference between two object graphs they become a single object graph.

**Comment by oatsoda** [comment]
Ok, interesting, I did not appreciate this. 

I would have thought, however, that modelling business logic would raise instances where I would want to use the same Business Object in more than one place in the domain.  Are we expected therefore to create a Business Object for each situation that we use it?  Can we utilise inheritance to accomplish this - i.e. the leaf nodes for each situation?

Does the same apply to Child Lists?  I could have a BusinessListBase called AppointmentList and potentially use this on several different Business Objects as a child list to save different criteria of Appointments.

**Comment by jonnybee** [comment]
Think instance here. One - 1 child object instance can only be owned by One - 1 parent. 

You may use the same type as child object - bu you cannot use the same instance in multiple parents or lists.

**Comment by oatsoda** [comment]
OK, so by instance you mean "unique entity"? 

For example, Company (parent) can have an Address (child) and Contact (parent) can have and Address (child) - but you are saying that the child Address should never be the same unique instance for both the Company and Contact - they should not share the address.

Are we really talking here about managing concurrency?

In the above example, as long as I ensure the Company and Contact objects are edited in isolation of each other then there aren't any physical CSLA constraints that will stop me from doing this?

**Comment by jonnybee** [comment]
This is where talk about:
 
- owning reference 
- inter graph reference
- using reference 

These are documented in the Using CSLA4 Creating Business Objects ebook.

Owning reference is the default where a "parent" owns a "child" object. 
 
What you describe here is the "inter graph" reference. 

The point here is that when you edit fields (and save) the owning reference will make CSLA serialize and deserialize objects. This happends for each edit of fields (from the BeginEdit/EndEdit/CancelEdit sequence and  support N-level undo), the Clone mechanism (which is also called before Save) and possible other events. 

When you have intergraph reference the object is "owned" by another object and should not be serialized/deserialied into a "new" instance in memory when editing or saving!  

So when you design objects it is *very* important to understand these reference types and use these appropriately.

**Comment by oatsoda** [comment]
Thanks Jonny.

I have the ebooks and I have read and re-read this section and never quite grasped it fully.

In my example I have two parents who share the same Child - same by Type and by Unique Identity - when I save the Parent, I **require** the Child to be saved as well - for **both** Parents.

Am I describing inter-graph in my example?  In the ebook, it describes this as being a reference to a child object that is **already in the child graph** so therefore you do not want the "inter-graph" to be treated as *another* child object.

But as I say, I have never fully grasped this.

**Comment by oatsoda** [comment]
Reading some more I don't quite see how my example 
> Company (parent) can have an Address (child) and Contact (parent) can have and Address (child)

is an inter-graph reference.  
I suppose if I look at Rocky's comment 
> If you create a reference between two object graphs they become a single object graph.

then I assume what you are saying is that if Address is Child of Contact and Company then I need to   change the modelling and combine them into one object graph - at which point one Parent will have a Child ref and the other an Inter-graph ref.

Is this what you meant?

Does this same principle apply to objects that can be used as a Child **or** Root?  i.e. Is creating an object that can be used in Child **and** Root scenarios a bad idea?

**Comment by jonnybee** [comment]
Let's say that Company has a list of office addresses and Contacts have a property for address. This is a very typical inter-graph relationship where Company is the "owner" of the address list and Contach ha an inter graph reference.  

Do NOT confuse this with switchable object. Basically - a switchable object must implement both the "root" data portal and "child" data portal and depending on which method (ex DataPortal.Fetch() vs DataPortal.FetchChild()) is used to create the object, it is either a "toot" object of a data structure or a "child" object and part of another "root" objects data structure.

**Comment by oatsoda** [comment]
Yes, but it would mean that in my example I would need to ensure Company and Contact are on the **same object graph** to allow me to use an inter-graph reference.  My original example I was proposing that Company and Contact were **not** on the same object graph.

No, I'm not confusing "switchable" with inter-graph.  My second comment was going back to my original problem where if I serialized a Parent I would serialize the entire graph.  This led to the discussion around Child objects being used in mulitple object graphs - which leads to difficulties in managing cached objects.  I am now saying that even if I am careful to follow what Rocky said about not using Child instance in more than one object graph, I think that I will then have issues with "switchable" objects because they could be cached as a Child in one place, but then updated as a Root in another place.  I would therefore have to manage Cache Invalidation across different object graphs.
